### PR TITLE
Collision Detector 2 © - 🧱🏃‍♂️🔥

### PR DESCRIPTION
**Pull request by Peter-JanGootzen**
# Collision Wars - The Collision Strikes Back
Very epic new collision system - Collision Detector 2 ©
## DisplacementSystem
The new collisions system will check if you are colliding and the displacement system will displace the entity back to where it should be (before the collision). The calculations are based on [this resource](https://github.com/noonat/intersect).
## Tagging exceptions
When an entity is a trigger they will fall through everything, EXCEPT the given tags you want it to collide with. For example the weapon and critter both have an exception tag "Platform" so they fall through everything except platform.

# How can this be tested?

Test with [46](https://github.com/Brick-Studios/BeastArena/pull/46)

Try to collide with as many things in as many different ways as possible. And focus on hitting things in the corners, because if the velocity is fast and the direction in a slighty off tilted up angle, some weird displacing of the entities could happen (we have not been able to spot this).

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] Any dependent changes have been merged and published in downstream modules
- [x] Code is const correct

WhatsApp Image 2019-11-30 at 22 58 20
WhatsApp Image 2019-11-30 at 23 22 14
Capture

### CODE REVIEW

File: include/brickengine/collision_detector_2.hpp [hunk 1]
```diff
@@ -0,0 +1,97 @@
+#ifndef FILE_COLLISION_DETECTOR_2_HPP
```
Comment (LuukSantegoeds) [inline include/brickengine/collision_detector_2.hpp]: If you call it "collision detector 2", other people might go out looking for 1

File: include/brickengine/std/sign.hpp [hunk 2]
```diff
@@ -0,0 +1,8 @@
+#ifndef FILE_SIGN_HPP
+#define FILE_SIGN_HPP
+
+template <typename T> int sign(T val) {
+    return (T(0) < val) - (val < T(0));
```
Comment (markJavab) [inline include/brickengine/std/sign.hpp]: I would add a comment what this does. I reckon it is the following but would go in a little more detail; https://stackoverflow.com/questions/1903954/is-there-a-standard-sign-function-signum-sgn-in-c-c

File: lib/collision_detector_2.cpp [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+        if (is_trigger_tag_exceptions.count(tag_2)) {
+            for (const std::string& tag_1 : tags_1) {
+                if (is_trigger_tag_exceptions.at(tag_2).count(tag_1)) {
+                    return true;
+                }
+            }
+        }
+    }
+    return false;
+}
```
Comment (markJavab) [inline lib/collision_detector_2.cpp]: I would add some comments about the functionality of this method. I do not know what it does, as it is called 'find', but it returns a boolean when looking for exceptions at two tags?

Comment (markJavab) [inline lib/collision_detector_2.cpp]: I would change the name of this method and add some comments to make it more clear what is happening here.  The name of the method is find for instance, but it returns a boolean

File: lib/collision_detector_2.cpp [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+        // Can't collide with yourself
+        if (opposite_id == entity_id) continue;
+        auto [ opposite_position, opposite_scale ] = em.getAbsoluteTransform(opposite_id);
+        // Can't collide with family
+        if (entity_parent && *entity_parent == opposite_id) continue;
+        if (entity_children.count(opposite_id)) continue;
+
+        bool is_trigger = false;
+        if (entity_collider->is_trigger || opposite_collider->is_trigger) {
+            // If there are no is_trigger_exceptions for these colliding entities,
```
Comment (markJavab) [inline lib/collision_detector_2.cpp]: Are these like actual exceptions, so an egg cannot collide with chickens for instance? Not a stackoverflowexception?

File: lib/systems/physics_system.cpp [hunk 5]
```diff
@@ -1,40 +1,46 @@
...
-    for(auto [entityId, physics] : entitiesWithPhysics){
+    for(auto [entity_id, physics] : entities_with_physics){
         if (physics->kinematic != Kinematic::IS_NOT_KINEMATIC) {
             physics->vx = 0;
             physics->vy = 0;
             continue;
         }
 
-        auto transform = entityManager->getComponent<TransformComponent>(entityId);
-        // TODO make it so trigger goes through everything instead of only checking if player can move through trigger
```
[the remark this answers is not in the document]
Comment (JanSchollaert) [reply lib/systems/physics_system.cpp]: I removed this comment wtf, yeah it works now

File: include/brickengine/std/sign.hpp (hunk 2 above)
Comment (JanSchollaert) [reply include/brickengine/std/sign.hpp]: https://en.wikipedia.org/wiki/Sign_function

File: lib/collision_detector_2.cpp (hunk 4 above)
Comment (JanSchollaert) [reply lib/collision_detector_2.cpp]: See comment on arena

File: include/brickengine/collision_detector_2.hpp [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+   - Collision Returns
+   - The Dark Collision
+   - Indiana Jones and the Temple of Collisions
+*/
+
+struct Collision {
+    Collision(int opposite_id, bool is_trigger) : opposite_id(opposite_id), is_trigger(is_trigger) {}
+    int opposite_id;
+    bool is_trigger;
+};
```
[the remark this answers is not in the document]
Comment (JanSchollaert) [reply include/brickengine/collision_detector_2.hpp]: These structs are all needed by the collision detector 2 so it makes sense having them here

File: lib/collision_detector_2.cpp [hunk 7]
```diff
@@ -176,6 +182,7 @@ ContinuousCollision CollisionDetector2::detectContinuousCollision(int entity_id,
     ContinuousCollision collision { opposite_opt, space_left_start_value };
 
     for (auto& [ opposite_id, opposite_collider ] : opposite_entities_with_collider) {
+        // TODO: If opposite_id = colliders creator_id (for bullets, when a player/weapon shoots, spawns a bullet, you should not collide with it)
```
Comment (LuukSantegoeds) [inline lib/collision_detector_2.cpp]: If this needs to get resolved, this is probably the best time to do so

Comment (Peter-JanGootzen) [reply lib/collision_detector_2.cpp]: This is an outdated comment, we have already fixed this by using should_displace = false in the collider of the bullet and damage_system not dealing damage to yourself.

**Pull request merged**